src/algebra/homology/local_cohomology.lean

```diff
@@ -0,0 +1,181 @@
+  applications*][brodmannsharp13]
+* [S. Iyengar, G. Leuschke, A. Leykin, Anton, C. Miller, E. Miller, A. Singh, U. Walther,
+  *Twenty-four hours of local cohomology*][iyengaretal13]
+
+## Tags
+
+local cohomology, local cohomology modules
+
+## Future work
+
+* Prove that this definition is equivalent to:
+    * the right-derived functor definition
+    * the characterization as the limit of Koszul homology
+    * the characterization as the cohomology of a Cech-like complex
+* Prove that local cohomology depends only on the radical of the ideal
+* Establish long exact sequence(s) in local cohomology
+-/
+
+open opposite
+open category_theory
+open category_theory.limits
+
+noncomputable theory
+
+universes u v
+
+/- We define local cohomology, implemented as a direct limit of Ext(R/J, --). -/
+namespace local_cohomology
+
+section
+variables {R : Type u} [comm_ring R] {D : Type v} [small_category D]
+
+/--  The directed system of `R`-modules of the form `R/J`, where `J` is an ideal of `R`,
+determined by the functor `I`  -/
+def ring_mod_ideals (I : D ⥤ ideal R) : D ⥤ Module.{u} R :=
+{ obj := λ t, Module.of R $ R ⧸ (I.obj t),
+  map := λ s t w, submodule.mapq _ _ (linear_map.id) (I.map w).down.down }
+
+/-- The diagram we will take the colimit of to define local cohomology, corresponding to the
+directed system determined by the functor `I` -/
+def diagram (I : D ⥤ ideal R) (i : ℕ) : Dᵒᵖ ⥤ Module.{u} R ⥤ Module.{u} R :=
+(ring_mod_ideals I).op ⋙ Ext R (Module.{u} R) i
+
+end
+
+section
+-- We momentarily need to work with a type inequality, as later we will take colimits
+-- along diagrams either in Type, or in the same universe as the ring, and we need to cover both.
+variables {R : Type max u v} [comm_ring R] {D : Type v} [small_category D]
+
+/-- `local_cohomology.of_diagram I i` is the the functor sending an module `M` over a commutative
+ring `R` to the direct limit of `Ext^i(R/J, M)`, where `J` ranges over a collection of ideals
+of `R`, represented as a functor `I`. -/
+/-
+In this definition we do not assume any special property of the diagram `I`, but the relevant case
+will be where `I` is (cofinal with) the diagram of powers of a single given ideal.
+
+Below, we give two equivalent (to be shown) definitions of the usual local cohomology with support
+in an ideal `J`, `local_cohomology` and `local_cohomology.of_radical_eq_radical`.
+
+TODO: Show that any functor cofinal with `I` gives the same result.
+ -/
+def of_diagram (I : D ⥤ ideal R) (i : ℕ) :
+  Module.{max u v} R ⥤ Module.{max u v} R :=
+colimit (diagram.{(max u v) v} I i)
+
+end
+
+section diagrams
+
+variables {R : Type u} [comm_ring R]
+
+/-- The functor sending a natural number `i` to the `i`-th power of the ideal `J` -/
+def ideal_powers (J : ideal R) : ℕᵒᵖ ⥤ ideal R :=
+{ obj := λ t, J^(unop t),
+  map := λ s t w, ⟨⟨ideal.pow_le_pow w.unop.down.down⟩⟩, }
+
+/-- The full subcategory of all ideals with the same radical as a given ideal `J` -/
```

Comment:
The docstring says that this is the type of ideals with the same radical as J, but the definition just uses an inequality. Is the docstring+name (or def) wrong?

Reply:
We discussed this offline. Summary:
- @adamtopaz pointed out that the condition might as well be `J ≤ J'.radical` rather than `J.radical ≤ J'.radical` (equivalent).
- Changed the name to `self_le_radical` accordingly. Fixed the docstrings.

The purpose of this diagram is to be used in showing (in a future PR) that local cohomology only depends on the ideal up to radical, and to give the loosest possible setting in which to show that some given diagram of ideals is cofinal with the standard one.